job-runner/chain/traverser.go

```diff
@@ -377,3 +395,85 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
 		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
 	}
 }
+
+// retrySequence retries a sequence by doing the following
+//   1. Mark the failed job and all jobs before it in the sequence to PENDING
+//   state, so that they are ran again on sequence retry. This is OK to do
+//   since all jobs are idempotent.
+//   2. Decrement the number of jobs that were previously ran in the sequence
+//   from the chain job count.
+//   3. Increment retry count for the sequence
+//   4. Enqueue the first job in the sequence to be ran again. It'll
+//   subsequently enqueue the remaining jobs in the sequence.
+func (t *traverser) retrySequence(failedJob proto.Job) error {
+	sequenceStartJob := t.chain.SequenceStartJob(failedJob)
+
+	jobIdsToRetry := t.jobIdsToRetry(sequenceStartJob.Id)
+	// jobIdsToRetry is now a list of jobs we have to undo
+	for _, jobId := range jobIdsToRetry {
+		t.chain.SetJobState(jobId, proto.STATE_PENDING)
+	}
+
+	// Decrement number of jobs we will attempt to retry
+	t.chain.n -= uint(len(jobIdsToRetry))
+
+	// Increment retry count for this sequence
+	t.chain.IncrementSequenceRetryCount(failedJob)
+
+	// Enqueue first job in sequence
+	t.runJobChan <- sequenceStartJob
+
+	return nil
+}
+
+// jobIdsToRetry does a breadth first search to find all jobs whose state is not PENDING, which means the state is either COMPLETE or FAILED. This is a list of jobs that need to be retried.
+func (t *traverser) jobIdsToRetry(sequenceStartJobId string) []string {
```

Comment:
Little difficult to follow this function, but if I'm reading it correctly and we have:
`A > [B > C] > D > E`
where `[ ]` denotes a sequence, and `C` fails, then `sequenceStartJobId=B` and we walk the graph all the way to `E`?   If true, we're doing too much work because `D` and `E` don't need to be reset.

In the previous incarnation of this algo (the simpler one in `retrySequence()` we stopped walking at `C`.

Reply:
Granted, it's late (11:30pm) my time, and I'm new to this code, but I cannot follow this function even though I know its intention. We should do a code walkthrough through with @BrianIp since he did similar code in grapher. For example, what boggles me is that we loop on `frontier` but in that loop we both add to _and_ delete from `frontier`.